python/ray/util/spark/start_ray_node.py

```diff
@@ -54,63 +54,42 @@
-                # "ray start ..." command process is still alive. Force to kill it.
-                process.kill()
-
-            # Release the shared lock, representing current ray node does not use the
-            # temp dir.
-            fcntl.flock(lock_fd, fcntl.LOCK_UN)
-
-            try:
-                # acquiring exclusive lock to ensure copy logs and removing dir safely.
-                fcntl.flock(lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
-                lock_acquired = True
-            except BlockingIOError:
-                # The file has active shared lock or exclusive lock, representing there
-                # are other ray nodes running, or other node running cleanup temp-dir
-                # routine. skip cleaning temp-dir, and skip copy logs to destination
-                # directory as well.
-                lock_acquired = False
-
-            if lock_acquired:
-                # This is the final terminated ray node on current spark worker,
-                # start copy logs (including all local ray nodes logs) to destination.
-                if collect_log_to_path:
-                    try:
-                        copy_log_dest_path = os.path.join(
-                            collect_log_to_path,
-                            os.path.basename(temp_dir) + "-logs",
-                            socket.gethostname(),
-                        )
-                        ray_session_dir = os.readlink(
-                            os.path.join(temp_dir, "session_latest")
-                        )
-                        shutil.copytree(
-                            os.path.join(ray_session_dir, "logs"),
-                            copy_log_dest_path,
-                        )
-                    except Exception as e:
-                        _logger.warning(
-                            "Collect logs to destination directory failed, "
-                            f"error: {repr(e)}."
-                        )
-
-                # Start cleaning the temp-dir,
-                shutil.rmtree(temp_dir, ignore_errors=True)
-        except Exception:
-            # swallow any exception.
-            pass
-        finally:
-            fcntl.flock(lock_fd, fcntl.LOCK_UN)
-            os.close(lock_fd)
+        _try_clean_temp_dir_at_exit(
+            process=process,
+            collect_log_to_path=collect_log_to_path,
+            temp_dir=temp_dir,
+            lock_fd=lock_fd,
+        )
+
+    def check_parent_alive() -> None:
+        orig_parent_id = os.getppid()
+        while True:
+            time.sleep(0.5)
+            if os.getppid() != orig_parent_id:
+                process.terminate()
+                try_clean_temp_dir_at_exit()
+                os._exit(143)
+
+    # When global mode is enabled, parent process could be detached,
+    # so we don't kill the process when parent process died.
+    if not global_mode_enabled():
+        threading.Thread(target=check_parent_alive, daemon=True).start()
```

Comment:
issue:

For global mode, we still need to enable `check_parent_alive` thread for **ray worker process** .Because we cancel ray worker nodes by cancelling spark job tasks, we have to rely on  `check_parent_alive` to make ray worker node killed

Reply:
Isn't the worker processes shutdown controlled by `_cancel_background_spark_job`?

Reply:
_cancel_background_spark_job only cancel spark job , so that all pyspark worker processes are killed, but the subprocesses of pyspark workers are not killed :)